runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -42,120 +59,211 @@ public final class LogReplicationUtils {
+                        } else {
+                            log.error("Unexpected type of Transaction Aborted Exception", tae);
+                        }
+                        throw new RetryNeededException();
+                    } catch (Exception e) {
+                        log.error("Unexpected exception type hit", e);
+                        throw new RetryNeededException();
                     }
-                    throw new RetryNeededException();
-                } catch (Exception e) {
-                    log.error("Unexpected exception type hit", e);
-                    throw new RetryNeededException();
-                }
-            }).run();
-        } catch (InterruptedException e) {
-            throw new StreamingException(e, StreamingException.ExceptionCause.SUBSCRIBE_ERROR);
-        } finally {
-            MicroMeterUtils.time(subscribeTimer, "logreplication.subscribe.duration");
+                }).setOptions(retrySettings).run();
+            } catch (InterruptedException e) {
+                throw new StreamingException(e, StreamingException.ExceptionCause.SUBSCRIBE_ERROR);
+                // TODO pankti: Unsubscribe the listener
+            } catch (RetryExhaustedException re) {
+                listener.onError(new SubscribeMaxRetryException());
+            }
+
+            if (!Objects.isNull(fullSyncTimestamp)) {
+                // Subscribe from the snapshot timestamp of this transaction, i.e., log tail when the transaction started.
+                // Subscribing from the commit address will result in missed updates which took place between the start
+                // and end of the transaction because reads in a transaction observe updates only till the snapshot when it
+                // started.
+                corfuStore.getRuntime().getTableRegistry().getStreamingManager().subscribeLogReplicationListener(
+                        listener, nsToTableName, nsToStreamTags, fullSyncTimestamp, bufferSize);
+            } else {
+                // We should never reach here with unsubscribe/resubscribe implementation.
+                log.warn("Unable to subscribe listener!");
+            }
+            return null;
         }
     }
 
+    public static void attemptClientFullSync(LogReplicationListener listener, CorfuStore corfuStore,
+                                             String namespace, Map<String, String> nsToStreamTags,
+                                             Map<String, List<String>> nsToTableName, int bufferSize,
+                                             ExecutorService clientExecutorService) {
+        clientExecutorService.submit(new ClientFullSyncAndSubscriptionTask(listener, corfuStore,
+                namespace, nsToStreamTags, nsToTableName, bufferSize));
+    }
 
     private static Table<LogReplicationSession, ReplicationStatus, Message> openReplicationStatusTable(CorfuStore corfuStore) {
```

Comment:
minor: this method can now be a non-static method inside ClientFullSyncAndSubscriptionTask.